versiondb/extsort/sort.go

```diff
@@ -0,0 +1,235 @@
+// It split the inputs into chunks, sort each chunk separately and save to disk file,
+// then provide an iterator of sorted items by doing a k-way merge with all the sorted chunk files.
+type ExtSorter struct {
+	// directory to store temporary chunk files
+	tmpDir string
+	// target chunk size
+	chunkSize  int64
+	lesserFunc LesserFunc
+
+	// current chunk
+	currentChunk     [][]byte
+	currentChunkSize int64
+
+	// manage the chunk goroutines
+	chunkWG sync.WaitGroup
+	lock    sync.Mutex
+	// finished chunk files
+	chunkFiles []*os.File
+	// chunking goroutine failure messages
+	failures []string
+}
+
+// New creates a new `ExtSorter`.
+func New(tmpDir string, chunkSize int64, lesserFunc LesserFunc) *ExtSorter {
+	return &ExtSorter{
+		tmpDir:     tmpDir,
+		chunkSize:  chunkSize,
+		lesserFunc: lesserFunc,
+	}
+}
+
+// Feed add un-ordered items to the sorter.
+func (s *ExtSorter) Feed(item []byte) error {
+	if len(item) > math.MaxUint32 {
+		return errors.New("item length overflows uint32")
+	}
+
+	s.currentChunkSize += int64(len(item)) + 4
+	s.currentChunk = append(s.currentChunk, item)
+
+	if s.currentChunkSize >= s.chunkSize {
+		return s.sortChunkAndRotate()
+	}
+	return nil
+}
+
+// sortChunkAndRotate sort the current chunk and save to disk.
+func (s *ExtSorter) sortChunkAndRotate() error {
+	chunkFile, err := os.CreateTemp(s.tmpDir, "sort-chunk-*")
+	if err != nil {
+		return err
+	}
+
+	// rotate chunk
+	chunk := s.currentChunk
+	s.currentChunk = nil
+	s.currentChunkSize = 0
+
+	s.chunkWG.Add(1)
+	go func() {
+		defer s.chunkWG.Done()
+		if err := sortAndSaveChunk(chunk, s.lesserFunc, chunkFile); err != nil {
+			chunkFile.Close()
+			s.lock.Lock()
+			defer s.lock.Unlock()
+			s.failures = append(s.failures, err.Error())
+			return
+		}
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		s.chunkFiles = append(s.chunkFiles, chunkFile)
+	}()
+	return nil
+}
+
+// Finalize wait for all chunking goroutines to finish, and return the merged sorted stream.
+func (s *ExtSorter) Finalize() (*MultiWayMerge, error) {
+	// handle the pending chunk
+	if s.currentChunkSize > 0 {
+		if err := s.sortChunkAndRotate(); err != nil {
+			return nil, err
+		}
+	}
+
+	s.chunkWG.Wait()
+	if len(s.failures) > 0 {
+		return nil, errors.New(strings.Join(s.failures, "\n"))
+	}
+
+	streams := make([]NextFunc, len(s.chunkFiles))
+	for i, chunkFile := range s.chunkFiles {
+		if _, err := chunkFile.Seek(0, 0); err != nil {
+			return nil, err
+		}
+		var lastItem []byte
+		reader := snappy.NewReader(chunkFile)
+		streams[i] = func() ([]byte, error) {
+			item, err := readItem(reader, lastItem)
+			if err == io.EOF {
+				return nil, nil
+			}
+			lastItem = item
+			return item, err
+		}
+	}
+
+	return NewMultiWayMerge(streams, s.lesserFunc)
+}
+
+// Close closes and remove all the temporary chunk files
+func (s *ExtSorter) Close() error {
+	var err error
+	for _, chunkFile := range s.chunkFiles {
+		err1 := chunkFile.Close()
+		if err != nil {
```

Comment:
should we append the error instead of assigning it if there is more than one close error?

Reply:
not sure if it worth the trouble, I'll try to find a library to group errors if there's any.

Reply:
done, `hashicorp/go-multierror` is good at this